go-controller/pkg/node/gateway_localnet.go

```diff
@@ -239,88 +246,357 @@ func localnetIPTablesHelper(subnet *net.IPNet) (util.IPTablesHelper, error) {
-			continue
+			return nil, err
+		}
+		socket = listener
+	case kapi.ProtocolUDP:
+		addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf(":%d", port))
+		if err != nil {
+			return nil, err
+		}
+		conn, err := net.ListenUDP("udp", addr)
+		if err != nil {
+			return nil, err
 		}
+		socket = conn
+	default:
+		return nil, fmt.Errorf("unknown protocol %q", protocol)
+	}
+	return socket, nil
+}
 
-		nodePort := fmt.Sprintf("%d", svcPort.NodePort)
-		rules = append(rules, iptRule{
+type localnetNodePortWatcherData struct {
+	recorder         record.EventRecorder
+	ipt              util.IPTablesHelper
+	gatewayIP        string
+	gatewayNextHopIP string
+	localAddrSet     map[string]net.IPNet
+}
+
+func (npw *localnetNodePortWatcherData) getLocalAddrs() error {
+	npw.localAddrSet = make(map[string]net.IPNet)
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return err
+	}
+	for _, addr := range addrs {
+		ip, ipNet, err := net.ParseCIDR(addr.String())
+		if err != nil {
+			return err
+		}
+		npw.localAddrSet[ip.String()] = *ipNet
+	}
+	klog.V(5).Infof("node local addresses initialized to: %v", npw.localAddrSet)
+	return nil
+}
+
+func (npw *localnetNodePortWatcherData) nodePortIPTRules(svcPort v1.ServicePort, gatewayIP string) []iptRule {
+	return []iptRule{
+		iptRule{
 			table: "nat",
 			chain: iptableNodePortChain,
 			args: []string{
-				"-p", string(protocol), "--dport", nodePort,
-				"-j", "DNAT", "--to-destination", net.JoinHostPort(gatewayIP, nodePort),
+				"-p", string(svcPort.Protocol),
+				"--dport", fmt.Sprintf("%d", svcPort.NodePort),
+				"-j", "DNAT",
+				"--to-destination", net.JoinHostPort(gatewayIP, fmt.Sprintf("%d", svcPort.NodePort)),
 			},
-		})
-		rules = append(rules, iptRule{
+		},
+		iptRule{
 			table: "filter",
 			chain: iptableNodePortChain,
 			args: []string{
-				"-p", string(protocol), "--dport", nodePort,
+				"-p", string(svcPort.Protocol),
+				"--dport", fmt.Sprintf("%d", svcPort.NodePort),
 				"-j", "ACCEPT",
 			},
-		})
+		},
 	}
-	return rules
 }
 
-type localnetNodePortWatcherData struct {
-	ipt       util.IPTablesHelper
-	gatewayIP string
+func (npw *localnetNodePortWatcherData) externalIPTRules(svcPort v1.ServicePort, externalIP, dstIP string) []iptRule {
+	return []iptRule{
+		iptRule{
+			table: "nat",
+			chain: iptableExternalIPChain,
+			args: []string{
+				"-p", string(svcPort.Protocol),
+				"-d", externalIP,
+				"--dport", fmt.Sprintf("%v", svcPort.Port),
+				"-j", "DNAT",
+				"--to-destination", net.JoinHostPort(dstIP, fmt.Sprintf("%v", svcPort.Port)),
+			},
+		},
+		iptRule{
+			table: "filter",
+			chain: iptableNodePortChain,
+			args: []string{
+				"-p", string(svcPort.Protocol),
+				"-d", externalIP,
+				"--dport", fmt.Sprintf("%v", svcPort.Port),
+				"-j", "ACCEPT",
+			},
+		},
+	}
+}
+
+func (npw *localnetNodePortWatcherData) isNetworkConnected(ipNet net.IP) bool {
+	for _, net := range npw.localAddrSet {
```

Comment:
I think you're right.

Reply:
Or maybe not come to think of it. `localAddrSet` will contain the IP address of every interface on the node as its key and that interface's `net.IPNet` as its value. It might just be that the `ExternalIP` provided is not exactly that IP address, but an IP address reachable on that subnet...hence we should perform the `net.Contains`

Reply:
ok